spec/lang/representation.md

```diff
@@ -405,22 +395,174 @@ Unsized types do not have values and thus there is no representation relation.
+                for (val, (_, ty)) in vals.zip(fields) {
+                    self.check_value(val, ty)?;
+                }
+            }
+            (Value::Tuple(vals), Type::Array { elem, count }) => {
+                ensure_ub(vals.len() == count, "Value::Tuple: invalid number of elements")?;
+                for val in vals {
+                    self.check_value(val, elem)?;
+                }
+            }
+            (Value::Union(chunk_data), Type::Union { chunks, .. }) => {
+                ensure_ub(chunk_data.len() == chunks.len(), "Value::Union: invalid chunk size")?;
+                for (data, (_, size)) in chunk_data.zip(chunks) {
+                    ensure_ub(data.len() == size.bytes(), "Value::Union: invalid chunk data")?;
+                }
+            }
+            (Value::Variant { discriminant, data }, Type::Enum { variants, .. }) => {
+                let Some(variant) = variants.get(discriminant) else {
+                    throw_ub!("Value::Variant: invalid discrimant type");
+                };
+                self.check_value(data, variant.ty)?;
+            }
+            (_, Type::Slice { .. }) => throw_ub!("Value: slices cannot be represented as values"),
+            _ => throw_ub!("Value: value does not match type")
+        }
+
+        ret(())
+    }
+}
+```
+
+## Typed memory accesses
+
+One key use of the value representation is to define a "typed" interface to memory.
+This interface is inspired by [Cerberus](https://www.cl.cam.ac.uk/~pes20/cerberus/).
+We also lift retagging from pointers to compound values (TODO: should this be here?).
+
+```rust
+impl<M: Memory> Machine<M> {
+    fn typed_store(&mut self, ptr: ThinPointer<M::Provenance>, val: Value<M>, ty: Type, align: Align, atomicity: Atomicity) -> Result {
+        // All values floating around in minirust must be valid.
+        assert!(self.check_value(val, ty).is_ok(), "trying to store {val:?} which is ill-formed for {:#?}", ty);
+        let bytes = ty.encode::<M>(val);
+        self.mem.store(ptr, bytes, align, atomicity)?;
+
+        ret(())
+    }
+
+    fn typed_load(&mut self, ptr: ThinPointer<M::Provenance>, ty: Type, align: Align, atomicity: Atomicity) -> Result<Value<M>> {
+        let bytes = self.mem.load(ptr, ty.size::<M::T>().expect_sized("the callers ensure `ty` is sized"), align, atomicity)?;
+        ret(match ty.decode::<M>(bytes) {
+            Some(val) => {
+                // Ensures we only produce valid values.
+                self.check_value(val, ty)?;
+                val
+            }
+            None => throw_ub!("load at type {ty:?} but the data in memory violates the validity invariant"), // FIXME use Display instead of Debug for `ty`
+        })
+    }
+}
+
+impl<M: Memory> ConcurrentMemory<M> {
+    /// Find all pointers in this value, ensure they are valid, and retag them.
+    fn retag_val(&mut self, frame_extra: &mut M::FrameExtra, val: Value<M>, ty: Type, fn_entry: bool) -> Result<Value<M>> {
```

Comment:
This should probably be moved to the `Validating a value` section in `statements.md`.